### PR TITLE
fix: media alt-text limit + nullability

### PR DESCRIPTION
**Pull request by llun**
## What

- Mastodon allows 1,500-character media descriptions (alt text); we stored them in a `varchar(255)` column, so >255-char alt text 500'd on PostgreSQL at upload (`MediaSchema` had no length cap) and was rejected with 422 on `PUT /api/v1/media/:id` (`.max(255)`).
- New migration widens `medias.description` to `text`; **both** reference schema dumps (`migrations/schema.sql`, `migrations/schema.sqlite.sql`) regenerated canonically per AGENTS.md.
- Upload (`POST /api/v1/media`, `POST /api/v2/media`) and update (`PUT`/`PATCH /api/v1/media/:id`) now share one `MAX_MEDIA_DESCRIPTION_LENGTH = 1500` cap; >1500 returns 422 like Mastodon.
- MediaAttachment serialization now emits `description: null` when no alt text is set (previously `''`), matching Mastodon's serializer; the S3 presigned-upload output and the web client's presigned-completion path were updated to match.
- `docs/features.md` gains one Planned Features checkbox for the remaining MediaAttachment gaps: `blurhash` computation and `gifv` detection.

## Tests

- Schema-level `it.each` limit tests (1500 vs 1501), upload-route 200/422 boundary tests, update-route boundary tests, and null-description serializer tests; existing `''`-expectation tests updated to `null`.

## Notes

- Migration + both schema dumps regenerated against fresh local DBs (SQLite `.schema`, throwaway local PostgreSQL 17 `pg_dump`). SQLite dump: `medias` table rebuilt with `description text` (all 3 indexes preserved). PostgreSQL dump: single-line `varchar(255)` → `text`.
- Part of Phase 2 (Entity & Parameter Gaps), PR 2.2. Independent of the other Phase 2 PRs.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/services/medias/types.ts [hunk 1]
```diff
@@ -48,7 +52,9 @@ export type FocusSchema = z.infer<typeof FocusSchema>
 export const MediaSchema = z.object({
   file: FileSchema,
   thumbnail: FileSchema.optional(),
-  description: z.string().optional(),
+  // Mastodon's alt-text limit. The column is `text`, so this cap is API
+  // compatibility, not a storage constraint.
+  description: z.string().max(MAX_MEDIA_DESCRIPTION_LENGTH).optional(),
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/services/medias/types.ts]: Good catch — done in 20f2f2e6. `MediaSchema.description` now normalizes empty/whitespace-only alt text to `null` on upload, matching `UpdateMediaRequest` and the AGENTS.md rule to keep create/update normalization consistent.

One adjustment to the exact suggestion: I kept `.optional()` **outermost** (`.string().max(...).transform(...).optional()`) rather than `.optional().nullable().transform(...)`. With the transform outside `.optional()`, the inferred output makes `description` a **required** key (`string | null`), which breaks callers that construct `MediaSchema` values as `{ file }` — `yarn build` failed with a type error. Placing `.optional()` last preserves the optional key while still coercing a present-but-blank value to `null`. Build + full test suite green.

File: lib/services/medias/types.test.ts [hunk 2]
```diff
@@ -14,3 +14,34 @@ describe('PresigedMediaInput', () => {
     expect(parsed.checksum).toBe('a9993e364706816aba3e25717850c26c9cd0d89d')
   })
 })
+
+describe('MediaSchema', () => {
```
Comment (llun) [inline lib/services/medias/types.test.ts:18]: 🤖 **Sub-agent review (Round 2) — test-coverage gap (medium), already addressed.**

The empty/whitespace-only → `null` transform added to `MediaSchema.description` (in response to the Round 1 review) was **unguarded**: this `MediaSchema` block only asserted the 1500/1501 `.max()` boundary, and both cases stay green whether or not the transform exists. So a regression that removed the normalization would not be caught.

Fixed in c3f2764a: the block now asserts parsed values — `` and `"   "` normalize to `null`, and a present 1500-char value passes through unchanged.

File: app/api/v1/media/[id]/route.ts [hunk 3]
```diff
@@ -41,17 +42,18 @@ interface Params {
...
 // from the raw payload so a partial update never clears the field the client
 // omitted.
 const UpdateMediaRequest = z.object({
   description: z
     .string()
-    .max(255)
+    .max(MAX_MEDIA_DESCRIPTION_LENGTH)
     .nullable()
     .optional()
     .transform((value) => (value && value.trim() ? value : null)),
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/media/[id]/route.ts]: Done in c3f2764a — moved `.optional()` outermost in `UpdateMediaRequest` so an omitted description parses to `undefined` rather than `null`, matching `MediaSchema`. No behavior change (the handler already gates on `'description' in payload`); build + full suite green.

File: lib/services/medias/types.ts [hunk 4]
```diff
@@ -48,7 +52,17 @@ export type FocusSchema = z.infer<typeof FocusSchema>
...
+  // compatibility, not a storage constraint. Empty/whitespace-only alt text is
+  // normalised to null so the upload path clears the description the same way
+  // the update path (UpdateMediaRequest) does. `.optional()` stays OUTERMOST so
+  // an omitted field remains an optional key (callers construct MediaSchema
+  // values without a description); a present-but-blank value becomes null.
+  description: z
+    .string()
+    .max(MAX_MEDIA_DESCRIPTION_LENGTH)
+    .transform((value) => (value && value.trim() ? value : null))
+    .optional(),
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/services/medias/types.ts:66]: Done in 361ec6395fcd6f064053ae68bc6fd862b4187207. Added `.nullable()` so `MediaSchema.description` is now identical to `UpdateMediaRequest`. In practice the sync upload path parses multipart form data (`Object.fromEntries(form.entries())`), whose values are always strings/Files rather than JS `null`, but this makes the two schemas structurally identical and handles an explicit null defensively. Added a null-input test case; build + full suite green.

File: lib/services/medias/getMediaAttachment.ts [hunk 5]
```diff
@@ -64,7 +64,9 @@ export const getMediaAttachment = (
         : {}),
       ...(media.focus ? { focus: media.focus } : {})
     },
-    description: media.description ?? '',
+    // Mastodon emits null (not '') when no alt text is set; '' from legacy rows
+    // is normalised to null too.
+    description: media.description || null,
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/services/medias/getMediaAttachment.ts]: Applied in 0a1e01a519cdb395c40bcc9d7393a5d9dc568610, with one correction to the suggested expression. Your `(media.description && media.description.trim()) || null` returns the **trimmed** value when a description has surrounding whitespace (e.g. `'  hello  '` → `'hello'`), which would silently strip content and diverge from the write-path transform (which preserves the original non-blank value). I used `media.description?.trim() ? media.description : null` instead: whitespace-only/empty → null, but a real description is returned **verbatim**. Added tests for both whitespace→null and the no-trim guarantee.

File: lib/services/medias/getMediaAttachment.test.ts [hunk 6]
```diff
@@ -74,10 +74,18 @@ describe('getMediaAttachment', () => {
...
+    expect(attachment.description).toBeNull()
+  })
+
+  it('normalises an empty stored description to null', () => {
+    const attachment = getMediaAttachment(
+      { ...baseMedia, description: '' },
+      'llun.test'
+    )
+    expect(attachment.description).toBeNull()
   })
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/services/medias/getMediaAttachment.test.ts:89]: Done in 0a1e01a519cdb395c40bcc9d7393a5d9dc568610 — added `normalises a whitespace-only stored description to null` plus a `returns a real description verbatim, without trimming` guard.

File: app/api/v1/media/[id]/route.ts [hunk 7]
```diff
@@ -41,20 +42,21 @@ interface Params {
...
     .string()
-    .max(255)
+    .max(MAX_MEDIA_DESCRIPTION_LENGTH)
     .nullable()
-    .optional()
-    .transform((value) => (value && value.trim() ? value : null)),
+    .transform((value) => (value && value.trim() ? value : null))
+    .optional(),
   focus: FocusSchema.optional()
 })
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/media/[id]/route.ts]: Great call — done in 9779499f0c40fd959c886d77464a5dbfa426a82c. `UpdateMediaRequest = MediaSchema.pick({ description: true, focus: true })` now, so the update and upload paths share one validation and cannot drift.

File: app/api/v1/media/[id]/route.ts [hunk 8]
```diff
@@ -8,6 +8,7 @@ import {
 import { headerHost } from '@/lib/services/guards/headerHost'
 import { AuthenticatedApiHandle } from '@/lib/services/guards/types'
 import { deleteMediaFile, saveMediaThumbnail } from '@/lib/services/medias'
+import { MAX_MEDIA_DESCRIPTION_LENGTH } from '@/lib/services/medias/constants'
 import { MediaValidationError } from '@/lib/services/medias/errors'
 import { getMediaAttachment } from '@/lib/services/medias/getMediaAttachment'
 import { FileSchema, FocusSchema } from '@/lib/services/medias/types'
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/media/[id]/route.ts]: Done in 9779499f0c40fd959c886d77464a5dbfa426a82c. Imports cleaned up: with the `.pick()` reuse, `FocusSchema`, `z`, and `MAX_MEDIA_DESCRIPTION_LENGTH` are all now unused, so I removed them (kept `FileSchema`, which the thumbnail path still uses) and added `MediaSchema`. Lint + build + full suite green.

**Comment by llun** [comment]
### 🤖 Review loop converged (6 rounds) — ready for maintainer review

Sub-agent review (5 dimensions + adversarial verification) each round, plus a gemini re-review each round:

- **R1** — gemini: normalize empty upload descriptions to null → fixed
- **R2** — sub-agent: the null-transform was untested (both max-boundary cases stay green without it) → strengthened tests; gemini: align `UpdateMediaRequest` `.optional()` ordering → fixed
- **R3** — gemini: add `.nullable()` so `MediaSchema` == `UpdateMediaRequest` → fixed
- **R4** — gemini: normalize whitespace-only legacy rows in the serializer (corrected to preserve non-blank values verbatim rather than trim content) → fixed
- **R5** — gemini: derive `UpdateMediaRequest` from `MediaSchema.pick({description, focus})` so the two paths can't drift → fixed
- **R6** — sub-agent: 0 findings; gemini: clean summary, no new threads → **converged**

Net effect: empty/whitespace/null alt text normalizes to `null` consistently across the upload, update, and serialization paths (now backed by one shared schema), all covered by tests. All review threads resolved; CI green (tests, build, lint/prettier, **Schema Dump Sync**, CodeQL).
